ksql-rest-app/src/test/java/io/confluent/ksql/rest/server/computation/RecoveryTest.java

```diff
@@ -0,0 +1,367 @@
+      fakeCommandQueue.getNewCommands().forEach(
+          c -> statementExecutor.handleStatement(
+              c.getCommand(), c.getCommandId(), Optional.empty())
+      );
+    }
+
+    void submitCommands(final String ...statements) {
+      for (final String statement : statements) {
+        final Response response = ksqlResource.handleKsqlStatements(
+            new KsqlRequest(statement, Collections.emptyMap()));
+        assertThat(response.getStatus(), equalTo(200));
+        executeCommands();
+      }
+    }
+  }
+
+  private void assertTopicsEqual(final KsqlTopic topic, final KsqlTopic otherTopic) {
+    assertThat(topic.getTopicName(), equalTo(otherTopic.getTopicName()));
+    assertThat(topic.getKafkaTopicName(), equalTo(otherTopic.getKafkaTopicName()));
+    assertThat(
+        topic.getKsqlTopicSerDe().getClass(),
+        equalTo(otherTopic.getKsqlTopicSerDe().getClass()));
+  }
+
+  private void assertTablesEqual(final KsqlTable table, final StructuredDataSource other) {
+    assertThat(other, instanceOf(KsqlTable.class));
+    final KsqlTable otherTable = (KsqlTable) other;
+    assertThat(table.isWindowed(), equalTo(otherTable.isWindowed()));
+  }
+
+  private void assertSourcesEqual(
+      final StructuredDataSource source,
+      final StructuredDataSource other) {
+    assertThat(source.getClass(), equalTo(other.getClass()));
+    assertThat(source.getName(), equalTo(other.getName()));
+    assertThat(source.getDataSourceType(), equalTo(other.getDataSourceType()));
+    assertThat(source.getSchema(), equalTo(other.getSchema()));
+    assertThat(source.getSqlExpression(), equalTo(other.getSqlExpression()));
+    assertThat(
+        source.getTimestampExtractionPolicy(), equalTo(other.getTimestampExtractionPolicy()));
+    assertThat(source, anyOf(instanceOf(KsqlStream.class), instanceOf(KsqlTable.class)));
+    assertTopicsEqual(source.getKsqlTopic(), other.getKsqlTopic());
+    if (source instanceof KsqlTable) {
+      assertTablesEqual((KsqlTable) source, other);
+    }
+  }
+
+  private void shouldRecover(final List<QueuedCommand> commands) {
+    // Given:
+    final KsqlServer executeServer = new KsqlServer(commands);
+    executeServer.executeCommands();
+    final KsqlEngine engine = executeServer.ksqlEngine;
+
+    // When:
+    final KsqlServer recoverServer = new KsqlServer(commands);
+    recoverServer.recover();
+    final KsqlEngine recovered = recoverServer.ksqlEngine;
+
+    // Then:
+    assertThat(
+        engine.getMetaStore().getAllStructuredDataSourceNames(),
+        equalTo(recovered.getMetaStore().getAllStructuredDataSourceNames()));
+    for (final String sourceName : engine.getMetaStore().getAllStructuredDataSourceNames()) {
+      assertSourcesEqual(
+          engine.getMetaStore().getSource(sourceName),
+          recovered.getMetaStore().getSource(sourceName)
+      );
+    }
+    final Collection<PersistentQueryMetadata> queries = engine.getPersistentQueries();
```

Comment:
Consider:

```
final Map<QueryId, PersistentQueryMetadata> queries = byId(engine.getPersistentQueries());
    final Map<QueryId, PersistentQueryMetadata> recoveredQueries = byId(recovered.getPersistentQueries());
    assertThat(recoveredQueries.keySet(), is(queries.keySet()));

    queries.forEach((queryId, query) -> {
          final PersistentQueryMetadata recovered = recoveredQueries.get(queryId);
          assertThat(recovered.getSourceNames(), equalTo(query.getSourceNames()));
          assertThat(recovered.getSinkNames(), equalTo(query.getSinkNames()));
          assertThat(recovered.getResultSchema(), equalTo(query.getResultSchema()));
          assertThat(recovered.getStatementString(), equalTo(query.getStatementString()));
        }
    );
 ```

As the assert on the collection sizes will now report _which_ query ids were missing.

Where `byId` is:
```
  private static Map<QueryId, PersistentQueryMetadata> byId(
      final Collection<PersistentQueryMetadata> queries) {
    return queries.stream()
        .collect(Collectors.toMap(PersistentQueryMetadata::getQueryId, Function.identity()));
  }
```